File: python/sentimental-credit/credit.py
# American Express uses 15-digit numbers, start with 34 or 37
# MasterCard uses 16-digit numbers, start with 51, 52, 53, 54, or 55
# Visa uses 13- and 16-digit numbers, start with 4.


card = input("Number: ")
copycard = card[::-1]

checksum = sum(int(x)*2 // 10 + int(x)*2 %
               10 for x in copycard[1::2]) + sum(int(x) for x in copycard[::2])

if checksum % 10 == 0:
    if len(card) == 15 and (int(card[0:2]) in [34, 37]):
        print("AMEX")
    elif len(card) == 16 and 50 < int(card[0:2]) < 56:
        print("MASTERCARD")
    elif len(card) in [13, 16] and int(card[0]) == 4:
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: c/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long cardno;
int sum;

int checksum(long cardno);
int card_len(long cardno);
int main(void)
{
    // take card input from user
    cardno = get_long("Number: ");
    int len = card_len(cardno);

    sum = checksum(cardno);

    // checking card name
    if (sum % 10 == 0)
    {
        // for 15 digit
        // querying first digit
        int check = cardno / 10000000000000;

        // for 16 digit
        // querying first digit of cardno
        int check1 = cardno / 1000000000000000;
        // querying first two digits of cardno
        int check2 = cardno / 100000000000000;

        // for 13 digit
        if (len == 13 && cardno / 1000000000000 == 4)
        {
            printf("VISA\n");
        }

        // for 15 digit
        else if (len == 15 && (check == 34 || check == 37))
        {
            printf("AMEX\n");
        }

        // for 16 digit
        else if (len == 16 && check1 == 4)
        {
            printf("VISA\n");
        }
        else if (len == 16 && (check2 >= 51 && check2 <= 55))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int card_len(long n)
{
    int len = 1;
    while (n > 9)
    {
        len++;
        n /= 10;
    }
    return len;
}

int checksum(long card)
{
    long total = 0;

    // digits at even places
    for (long i = 10; i < 100000000000000000; i = i * 100)
    {
        long n = card % (i * 10);
        n = n / i;
        n = n * 2;
        if (n > 9)
        {
            n = (n % 10) + 1;
        }
        total += n;
    }

    // digits at odd places
    for (long i = 1; i < 10000000000000000; i = i * 100)
    {
        long n = card % (i * 10);
        n = n / i;
        total += n;
    }

    return (int) total;
}
